zcash_test_vectors/transaction_zsa.py

```diff
@@ -0,0 +1,186 @@
+
+
+class AssetBurnDescription(object):
+    def __init__(self, rand):
+        isk = IssuanceKeys(rand.b(32))
+        desc_size = rand.u32() % 512 + 1
+        desc_bytes = get_random_unicode_bytes(desc_size, rand)
+        asset_digest_bytes = asset_digest(encode_asset_id(isk.ik, desc_bytes))
+        self.assetBase: Point = zsa_value_base(asset_digest_bytes)
+        self.valueBurn = rand.u64()
+
+    def __bytes__(self):
+        return bytes(self.assetBase) + struct.pack('<Q', self.valueBurn)
+
+
+class IssueActionDescription(object):
+    def __init__(self, rand, ik):
+        self.assetDescSize = rand.u32() % 512 + 1
+        self.asset_desc = get_random_unicode_bytes(self.assetDescSize, rand)
+        self.vNotes = []
+        for _ in range(rand.u8() % 5):
+            self.vNotes.append(IssueNote(rand, ik, self.asset_desc))
+        self.flagsIssuance = rand.u8() & 1  # Only one bit is reserved for the finalize flag currently
+
+    def __bytes__(self):
+        ret = b''
+
+        ret += write_compact_size(self.assetDescSize)
+        ret += bytes(self.asset_desc)
+        ret += write_compact_size(len(self.vNotes))
+        if len(self.vNotes) > 0:
+            for note in self.vNotes:
+                ret += bytes(note)
+        ret += struct.pack('B', self.flagsIssuance)
+
+        return ret
+
+
+class IssueNote(object):
+    def __init__(self, rand, ik, asset_desc):
+        fvk_r = FullViewingKey.from_spending_key(SpendingKey(rand.b(32)))
+        self.recipient = fvk_r.default_d() + bytes(fvk_r.default_pkd())
+        self.value = rand.u64()
+        asset_digest_bytes = asset_digest(encode_asset_id(ik, asset_desc))
+        self.assetBase = zsa_value_base(asset_digest_bytes)
+        self.rho = Point.rand(rand).extract()
+        self.rseed = rand.b(32)
+
+    def __bytes__(self):
+        ret = b''
+        ret += bytes(self.recipient)
+        ret += struct.pack('<Q', self.value)
+        ret += bytes(self.assetBase)
+        ret += bytes(self.rho)
+        ret += self.rseed
+
+        return ret
+
+
+class TransactionZSA(TransactionBase):
+    def __init__(self, rand, consensus_branch_id, have_orchard_zsa=True, have_burn=True, have_issuance=True):
+
+        # We cannot have burns without an OrchardZSA bundle.
+        assert have_orchard_zsa or not have_burn
+
+        # All Transparent, Sapling, and part of the Orchard Transaction Fields are initialized in the super class.
+        super().__init__(rand, have_orchard_zsa)
+
+        # Common Transaction Fields
+        self.nVersionGroupId = NU7_VERSION_GROUP_ID
+        self.nConsensusBranchId = consensus_branch_id
+
+        # OrchardZSA Transaction Fields
+        if have_orchard_zsa:
+            for _ in range(rand.u8() % 5):
+                self.vActionsOrchard.append(OrchardZSAActionDescription(rand))
+            self.flagsOrchard = rand.u8()
+            self.flagsOrchard = (self.flagsOrchard & 7) | 4  # Three flag bits are defined, we set enableZSA to true.
```

Comment:
Let's keep the comments always above the line (x2)

Reply:
Done.